File: R/draw_trial.R
# =============================================================================
# draw_trial.R — Compose a full Ebbinghaus stimulus from a trial row
# =============================================================================
# Takes a single trial row (one-row data frame or named list) and returns
# a complete ggplot object ready for display or saving.
#
# Consumes parameters from the trials table schema. Does not generate,
# modify, or infer any parameters.
# =============================================================================

source("Ebbinghaus/R/draw_shape.R")

#' Draw a complete Ebbinghaus stimulus from trial parameters
#'
#' @param trial A one-row data frame or named list with the trials table schema.
#'
#' @return A ggplot object (the complete stimulus image).
draw_trial <- function(trial) {


  # Coerce to list for easy access
  if (is.data.frame(trial)) {
    stopifnot(nrow(trial) == 1)
    trial <- as.list(trial)
  }

  # --- Compute center positions for groups A and B based on orientation ---
  positions <- compute_group_positions(trial$orientation, trial$canvas_width,
                                       trial$canvas_height)
  ax <- positions$a_x
  ay <- positions$a_y
  bx <- positions$b_x
  by <- positions$b_y

  # --- Compute bounding box for all shapes ---
  # We need to know the full extent of every shape so we can set axis limits
  # that contain everything without clipping.
  all_extents <- compute_all_extents(trial, positions)

  # --- Build the plot ---
  p <- ggplot() +
    theme_void() +
    theme(
      plot.background = element_rect(fill = trial$background_color, color = NA),
      plot.margin = margin(0, 0, 0, 0)
    )

  # --- Draw surround shapes (behind test shapes) ---
  # Group A surrounds
  if (!is.na(trial$surround_a_n) && trial$surround_a_n > 0) {
    surround_positions_a <- compute_surround_positions(
      cx = ax, cy = ay,
      n = trial$surround_a_n,
      distance = trial$surround_a_distance
    )
    for (j in seq_len(trial$surround_a_n)) {
      p <- p + draw_shape(
        shape = trial$surround_a_shape,
        x     = surround_positions_a$x[j],
        y     = surround_positions_a$y[j],
        size  = trial$surround_a_size,
        color = trial$surround_a_color,
        fill  = if (is.na(trial$surround_a_fill)) NA else trial$surround_a_fill
      )
    }
  }

  # Group B surrounds
  if (!is.na(trial$surround_b_n) && trial$surround_b_n > 0) {
    surround_positions_b <- compute_surround_positions(
      cx = bx, cy = by,
      n = trial$surround_b_n,
      distance = trial$surround_b_distance
    )
    for (j in seq_len(trial$surround_b_n)) {
      p <- p + draw_shape(
        shape = trial$surround_b_shape,
        x     = surround_positions_b$x[j],
        y     = surround_positions_b$y[j],
        size  = trial$surround_b_size,
        color = trial$surround_b_color,
        fill  = if (is.na(trial$surround_b_fill)) NA else trial$surround_b_fill
      )
    }
  }

  # --- Draw test shapes (on top) ---
  p <- p + draw_shape(
    shape = trial$test_a_shape,
    x     = ax,
    y     = ay,
    size  = trial$test_a_size,
    color = trial$test_a_color,
    fill  = if (is.na(trial$test_a_fill)) NA else trial$test_a_fill
  )

  p <- p + draw_shape(
    shape = trial$test_b_shape,
    x     = bx,
    y     = by,
    size  = trial$test_b_size,
    color = trial$test_b_color,
    fill  = if (is.na(trial$test_b_fill)) NA else trial$test_b_fill
  )

  # --- Set axis limits using coord_fixed ---
  # coord_fixed with explicit xlim/ylim clips visually at the boundary
  # but does NOT discard geometry (unlike scale xlim/ylim which remove data).
  # We use the computed bounding box to ensure all shapes are fully contained.
  p <- p +
    coord_fixed(
      xlim = c(all_extents$xmin, all_extents$xmax),
      ylim = c(all_extents$ymin, all_extents$ymax),
      clip = "off"
    )

  p
}


#' Compute center positions for groups A and B based on orientation
#'
#' @param orientation "horizontal", "vertical", or "diagonal"
#' @param canvas_width Canvas width
#' @param canvas_height Canvas height
#'
#' @return A list with a_x, a_y, b_x, b_y
compute_group_positions <- function(orientation, canvas_width, canvas_height) {

  cx <- canvas_width / 2
  cy <- canvas_height / 2

  # Offset: place groups at 25% and 75% of the relevant axis
  switch(orientation,
    horizontal = list(
      a_x = canvas_width * 0.25,  a_y = cy,
      b_x = canvas_width * 0.75,  b_y = cy
    ),
    vertical = list(
      a_x = cx,  a_y = canvas_height * 0.75,   # A = top
      b_x = cx,  b_y = canvas_height * 0.25    # B = bottom
    ),
    diagonal = list(
      a_x = canvas_width * 0.25,   a_y = canvas_height * 0.75,   # A = upper-left
      b_x = canvas_width * 0.75,   b_y = canvas_height * 0.25    # B = lower-right
    ),
    stop("Unknown orientation: ", orientation)
  )
}


#' Compute positions for N surround shapes equally spaced around a center
#'
#' @param cx Center x
#' @param cy Center y
#' @param n Number of surround shapes
#' @param distance Distance from center to each surround center
#'
#' @return A data frame with x and y columns
compute_surround_positions <- function(cx, cy, n, distance) {

  angles <- seq(0, 2 * pi, length.out = n + 1)[-(n + 1)]

  data.frame(
    x = cx + distance * cos(angles),
    y = cy + distance * sin(angles)
  )
}


#' Compute bounding box that contains all shapes in a trial
#'
#' Examines test shapes and surround shapes and returns the min/max x/y
#' needed to fully contain every shape, plus a padding margin.
#'
#' @param trial A trial as a list
#' @param positions Output of compute_group_positions()
#' @param padding Fraction of total extent to add as margin. Default 0.05 (5%).
#'
#' @return A list with xmin, xmax, ymin, ymax
compute_all_extents <- function(trial, positions, padding = 0.05) {

  # Start with test shape extents
  # For both circle (radius = size) and square (half-side = size),
  # the shape extends `size` units from center in each direction.
  all_x <- c(positions$a_x - trial$test_a_size, positions$a_x + trial$test_a_size,
              positions$b_x - trial$test_b_size, positions$b_x + trial$test_b_size)
  all_y <- c(positions$a_y - trial$test_a_size, positions$a_y + trial$test_a_size,
              positions$b_y - trial$test_b_size, positions$b_y + trial$test_b_size)

  # Add surround shape extents for group A
  if (!is.na(trial$surround_a_n) && trial$surround_a_n > 0) {
    spos_a <- compute_surround_positions(
      positions$a_x, positions$a_y,
      trial$surround_a_n, trial$surround_a_distance
    )
    all_x <- c(all_x, spos_a$x - trial$surround_a_size,
                       spos_a$x + trial$surround_a_size)
    all_y <- c(all_y, spos_a$y - trial$surround_a_size,
                       spos_a$y + trial$surround_a_size)
  }

  # Add surround shape extents for group B
  if (!is.na(trial$surround_b_n) && trial$surround_b_n > 0) {
    spos_b <- compute_surround_positions(
      positions$b_x, positions$b_y,
      trial$surround_b_n, trial$surround_b_distance
    )
    all_x <- c(all_x, spos_b$x - trial$surround_b_size,
                       spos_b$x + trial$surround_b_size)
    all_y <- c(all_y, spos_b$y - trial$surround_b_size,
                       spos_b$y + trial$surround_b_size)
  }

  xmin <- min(all_x)
  xmax <- max(all_x)
  ymin <- min(all_y)
  ymax <- max(all_y)

  # Add padding
  x_pad <- (xmax - xmin) * padding
  y_pad <- (ymax - ymin) * padding

  list(
    xmin = xmin - x_pad,
    xmax = xmax + x_pad,
    ymin = ymin - y_pad,
    ymax = ymax + y_pad
  )
}

File: py/src/draw_trial.py
"""
draw_trial.py — Compose a full Ebbinghaus stimulus from a trial row

Takes a single trial row (one-row DataFrame or dict) and returns
a matplotlib Figure ready for display or saving.
"""

import math
import numpy as np
import matplotlib.pyplot as plt

from src.draw_shape import draw_shape


def compute_group_positions(
    orientation: str, canvas_width: int, canvas_height: int
) -> dict:
    """Compute center positions for groups A and B based on orientation."""
    cx = canvas_width / 2
    cy = canvas_height / 2

    if orientation == "horizontal":
        return dict(
            a_x=canvas_width * 0.25, a_y=cy,
            b_x=canvas_width * 0.75, b_y=cy,
        )
    elif orientation == "vertical":
        return dict(
            a_x=cx, a_y=canvas_height * 0.75,
            b_x=cx, b_y=canvas_height * 0.25,
        )
    elif orientation == "diagonal":
        return dict(
            a_x=canvas_width * 0.25, a_y=canvas_height * 0.75,
            b_x=canvas_width * 0.75, b_y=canvas_height * 0.25,
        )
    else:
        raise ValueError(f"Unknown orientation: {orientation}")


def compute_surround_positions(
    cx: float, cy: float, n: int, distance: float
) -> list[tuple[float, float]]:
    """Compute positions for N surround shapes equally spaced around a center."""
    angles = np.linspace(0, 2 * np.pi, n, endpoint=False)
    return [(cx + distance * np.cos(a), cy + distance * np.sin(a)) for a in angles]


def compute_all_extents(trial: dict, positions: dict, padding: float = 0.05) -> dict:
    """Compute bounding box that contains all shapes in a trial."""
    all_x = [
        positions["a_x"] - trial["test_a_size"],
        positions["a_x"] + trial["test_a_size"],
        positions["b_x"] - trial["test_b_size"],
        positions["b_x"] + trial["test_b_size"],
    ]
    all_y = [
        positions["a_y"] - trial["test_a_size"],
        positions["a_y"] + trial["test_a_size"],
        positions["b_y"] - trial["test_b_size"],
        positions["b_y"] + trial["test_b_size"],
    ]

    sa_n = trial.get("surround_a_n", 0)
    if not _is_na(sa_n) and sa_n > 0:
        spos_a = compute_surround_positions(
            positions["a_x"], positions["a_y"],
            int(sa_n), trial["surround_a_distance"],
        )
        for sx, sy in spos_a:
            all_x.extend([sx - trial["surround_a_size"], sx + trial["surround_a_size"]])
            all_y.extend([sy - trial["surround_a_size"], sy + trial["surround_a_size"]])

    sb_n = trial.get("surround_b_n", 0)
    if not _is_na(sb_n) and sb_n > 0:
        spos_b = compute_surround_positions(
            positions["b_x"], positions["b_y"],
            int(sb_n), trial["surround_b_distance"],
        )
        for sx, sy in spos_b:
            all_x.extend([sx - trial["surround_b_size"], sx + trial["surround_b_size"]])
            all_y.extend([sy - trial["surround_b_size"], sy + trial["surround_b_size"]])

    xmin, xmax = min(all_x), max(all_x)
    ymin, ymax = min(all_y), max(all_y)
    x_pad = (xmax - xmin) * padding
    y_pad = (ymax - ymin) * padding

    return dict(
        xmin=xmin - x_pad, xmax=xmax + x_pad,
        ymin=ymin - y_pad, ymax=ymax + y_pad,
    )


def _is_na(val) -> bool:
    """Check if a value is NaN/None."""
    if val is None:
        return True
    try:
        return math.isnan(val)
    except (TypeError, ValueError):
        return False


def draw_trial(trial) -> plt.Figure:
    """Draw a complete Ebbinghaus stimulus from trial parameters.

    Args:
        trial: A one-row DataFrame or dict with the trials table schema.

    Returns:
        A matplotlib Figure.
    """
    if hasattr(trial, "iloc"):
        assert len(trial) == 1
        trial = trial.iloc[0].to_dict()

    positions = compute_group_positions(
        trial["orientation"], trial["canvas_width"], trial["canvas_height"]
    )
    extents = compute_all_extents(trial, positions)

    # Figure size in inches (96 dpi)
    w_in = trial["canvas_width"] / 96
    h_in = trial["canvas_height"] / 96
    fig, ax = plt.subplots(1, 1, figsize=(w_in, h_in))
    fig.patch.set_facecolor(trial["background_color"])
    ax.set_facecolor(trial["background_color"])

    ax.set_xlim(extents["xmin"], extents["xmax"])
    ax.set_ylim(extents["ymin"], extents["ymax"])
    ax.set_aspect("equal")
    ax.axis("off")

    # Draw surround shapes (behind test shapes)
    sa_n = trial.get("surround_a_n", 0)
    if not _is_na(sa_n) and sa_n > 0:
        spos_a = compute_surround_positions(
            positions["a_x"], positions["a_y"],
            int(sa_n), trial["surround_a_distance"],
        )
        for sx, sy in spos_a:
            draw_shape(
                ax, trial["surround_a_shape"], sx, sy,
                trial["surround_a_size"],
                color=trial["surround_a_color"],
                fill=trial["surround_a_fill"] if not _is_na(trial.get("surround_a_fill")) else None,
            )

    sb_n = trial.get("surround_b_n", 0)
    if not _is_na(sb_n) and sb_n > 0:
        spos_b = compute_surround_positions(
            positions["b_x"], positions["b_y"],
            int(sb_n), trial["surround_b_distance"],
        )
        for sx, sy in spos_b:
            draw_shape(
                ax, trial["surround_b_shape"], sx, sy,
                trial["surround_b_size"],
                color=trial["surround_b_color"],
                fill=trial["surround_b_fill"] if not _is_na(trial.get("surround_b_fill")) else None,
            )

    # Draw test shapes (on top)
    draw_shape(
        ax, trial["test_a_shape"],
        positions["a_x"], positions["a_y"],
        trial["test_a_size"],
        color=trial["test_a_color"],
        fill=trial["test_a_fill"] if not _is_na(trial.get("test_a_fill")) else None,
    )
    draw_shape(
        ax, trial["test_b_shape"],
        positions["b_x"], positions["b_y"],
        trial["test_b_size"],
        color=trial["test_b_color"],
        fill=trial["test_b_fill"] if not _is_na(trial.get("test_b_fill")) else None,
    )

    fig.subplots_adjust(left=0, right=1, top=1, bottom=0)
    return fig
